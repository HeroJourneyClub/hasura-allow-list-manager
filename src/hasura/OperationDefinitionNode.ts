import { OperationDefinitionNode } from 'graphql';
import { Source } from '@graphql-tools/utils';

export function getOperationDefinitionNodes(
  sources: Source[]
): OperationDefinitionNode[] {
  return sources.reduce<OperationDefinitionNode[]>((sourceAcc, source) => {
    return source.document.definitions.reduce((defAcc, def) => {
      if (def.kind == 'OperationDefinition') {
        return defAcc.concat(def)
      }
      return defAcc
    }, sourceAcc);
  }, [])
}
