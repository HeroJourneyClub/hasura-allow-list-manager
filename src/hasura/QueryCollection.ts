import { Source } from '@graphql-tools/utils';
import { DefinitionNode, FragmentDefinitionNode, OperationDefinitionNode, visit, print } from 'graphql';

export interface QueryCollection {
  name: string;
  query: string;
}


function addFragmentsToQuery(defNode: OperationDefinitionNode, fragments: Record<string, FragmentDefinitionNode>) {
  let query : string = print(defNode)

  function recursivelyFindFragments(defNode: OperationDefinitionNode| FragmentDefinitionNode) {
    visit(defNode, {
      FragmentSpread(node) {
        const fragDef = fragments[node.name.value]
        // everytime we find a fragment, we add it to the query collection, and check the fragment for fragment
        query += `\n\n${print(fragDef)}`
        recursivelyFindFragments(fragDef);
      },
    })
  }
  recursivelyFindFragments(defNode)

  return query
}

export function sourceToQueryCollection(prevQueryCollections: QueryCollection[], source: Source) {
    // Gather any fragment if any
    const fragmentMap = source.document.definitions.reduce<Record<string, FragmentDefinitionNode>>(
      (acc, nodeDef: DefinitionNode) => {
        if (nodeDef.kind == 'FragmentDefinition') {
          acc[nodeDef.name.value] = nodeDef
          return acc
        } else {
          return acc
        }
      },
      {}
    )

    return source.document.definitions.reduce<QueryCollection[]>(
      (acc, defNode: DefinitionNode) => {
        if (defNode.kind == 'OperationDefinition') {
          // Check if name already exist. If so raise an error
          const name = defNode.name.value
          const query = print(defNode)
          const exists = acc.find(qc => qc.name == name)
          if (exists) {
            if (exists.query != query) {
              throw Error(`Operation ${name} already exist. Please rename the operation name.`)
            } else {
              return acc
            }
          }

          return acc.concat({
            name: defNode.name.value,
            query: addFragmentsToQuery(defNode, fragmentMap),
          });
        } else {
          return acc
        }
      },
      prevQueryCollections
    );
  }


export function createQueryCollections(sources: Source[]): QueryCollection[] {
  return sources.reduce<QueryCollection[]>((acc, source) => {
    return sourceToQueryCollection(acc, source)
  }, []);
}

export function toMap(
  queries: QueryCollection[]
): Map<QueryCollection['name'], QueryCollection['query']> {
  return queries.reduce((acc, query) => {
    acc.set(query.name, query.query);
    return acc;
  }, new Map());
}

export function getAddedOrUpdatedQueries(
  oldQueries: QueryCollection[],
  newQueries: QueryCollection[]
) {
  const oldMap = toMap(oldQueries);

  return newQueries.reduce<{
    added: QueryCollection[];
    updated: QueryCollection[];
  }> (
    (acc, query) => {
      const oldQuery = oldMap.get(query.name);

      return {
        added: !!oldQuery ? acc['added'] : acc['added'].concat(query),
        updated: !!oldQuery && oldQuery !== query.query ? acc['updated'].concat(query) : acc['updated']
      };
    },
    { added: [], updated: [] }
  );
}
