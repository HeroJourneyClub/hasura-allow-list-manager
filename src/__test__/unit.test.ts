import * as path from 'path';
import { loadDocuments } from '@graphql-tools/load';
import { GraphQLFileLoader } from '@graphql-tools/graphql-file-loader';
import { createQueryCollections } from '../hasura';

async function getSource(pathStr: string) {
  return await loadDocuments(
    path.resolve(__dirname, pathStr),
    {
      loaders: [new GraphQLFileLoader()],
    }
  );
}

describe('createQueryCollections', () => {

  it('generate a query by operation', async () => {
    const sources = await getSource('operations/1/double.graphql')
    expect(sources.length).toBe(1)

    const queryCollections = createQueryCollections(sources);

    expect(queryCollections.length).toBe(2)
    expect(queryCollections).toMatchSnapshot()
  })

  it('handle fragment and move them to end of query', async () => {
    const sources = await getSource('operations/fragments/fragment.graphql')
    const queryCollections = createQueryCollections(sources);
    expect(queryCollections.length).toBe(1)
    expect(queryCollections).toMatchSnapshot()
  })

  it('handle recursive fragment and move them to end of query', async () => {
    const sources = await getSource('operations/fragments/recFragment.graphql')
    const queryCollections = createQueryCollections(sources);
    expect(queryCollections.length).toBe(1)
    expect(queryCollections).toMatchSnapshot()
  })

  it('ignores unused fragment', async () => {
    const sources = await getSource('operations/fragments/unusedFragment.graphql')
    const queryCollections = createQueryCollections(sources);
    expect(queryCollections.length).toBe(1)
    expect(queryCollections).toMatchSnapshot()
  })


  it('ignores other types than operation and fragment', async () => {
    const sources = await getSource('operations/unsupported.graphql')
    const queryCollections = createQueryCollections(sources);
    expect(queryCollections.length).toBe(1)
    expect(queryCollections).toMatchSnapshot()
  })
})
