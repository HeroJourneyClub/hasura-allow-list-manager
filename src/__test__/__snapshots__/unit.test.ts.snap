// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createQueryCollections generate a query by operation 1`] = `
Array [
  Object {
    "name": "firstQuery",
    "query": "query firstQuery {
  three {
    four
  }
}",
  },
  Object {
    "name": "secondQuery",
    "query": "query secondQuery {
  five {
    six
  }
}",
  },
]
`;

exports[`createQueryCollections handle fragment and move them to end of query 1`] = `
Array [
  Object {
    "name": "fragmentQuery",
    "query": "query fragmentQuery {
  testFragment {
    ...Frag
  }
  value
}

fragment Frag on FragType {
  fragValue
}",
  },
]
`;

exports[`createQueryCollections handle recursive fragment and move them to end of query 1`] = `
Array [
  Object {
    "name": "fragmentQuery2",
    "query": "query fragmentQuery2 {
  testFragment {
    ...Frag
  }
  value
}

fragment Frag on FragType {
  fragValue
  testFrag2 {
    ...Frag2
  }
}

fragment Frag2 on FragType {
  fragValue2
}",
  },
]
`;

exports[`createQueryCollections ignores other types than operation and fragment 1`] = `
Array [
  Object {
    "name": "fragmentQuery4",
    "query": "query fragmentQuery4 {
  testFragment {
    ...Frag
  }
  value
}

fragment Frag on FragType {
  fragValue
}",
  },
]
`;

exports[`createQueryCollections ignores unused fragment 1`] = `
Array [
  Object {
    "name": "fragmentQuery3",
    "query": "query fragmentQuery3 {
  value
  value2
}",
  },
]
`;
